src/lib/source-handlers/bitbucket-server/list-projects.ts

```diff
@@ -0,0 +1,87 @@
+import * as needle from 'needle';
+import * as debugLib from 'debug';
+import Bottleneck from 'bottleneck';
+import { BitbucketServerProjectData } from './types';
+import { getBitbucketServerToken } from './get-bitbucket-server-token';
+
+const debug = debugLib('snyk:bitbucket-server');
+
+const limiter = new Bottleneck({
+  reservoir: 1000, // initial value
+  reservoirRefreshAmount: 1000,
+  reservoirRefreshInterval: 3600 * 1000,
+  maxConcurrent: 1,
+  minTime: 1000,
+});
+
+limiter.on('failed', async (error, jobInfo) => {
+  const id = jobInfo.options.id;
+  console.warn(`Job ${id} failed: ${error}`);
+  if (jobInfo.retryCount === 0) {
+    // Here we only retry once
+    console.log(`Retrying job ${id} in 25ms!`);
+    return 25;
+  }
+});
+
+export const fetchAllProjects = async (
+  url: string,
+  token: string,
+): Promise<unknown[]> => {
+  let isLastPage = false;
+  let values: unknown[] = [];
+  let pageCount = 1;
+  let start = 0;
+  const limit = 100;
+  while (!isLastPage) {
+    debug(`Fetching page ${pageCount} for projects\n`);
+    const response = await limiter.schedule(() =>
+      needle(
+        'get',
+        `${url}/rest/api/1.0/projects?start=${start}&limit=${limit}`,
+        {
+          headers: { Authorization: 'Bearer ' + token },
+        },
+      ),
+    );
+    if (response.statusCode != 200) {
+      if (response.statusCode == 429) {
+        debug(
+          `Failed to fetch page: ${url}\n, Response Status: ${response.body}\nToo many requests \nWaiting for 3 minutes before resuming`,
+        );
+        await sleepNow(180000);
+        isLastPage = false;
+      } else {
+        debug(
+          `Failed to fetch page: ${url}\n, Response Status: ${response.statusCode}\nResponse Status Text: ${response.body} `,
+        );
+      }
+    }
+    const apiResponse = response.body['values'];
+    values = values.concat(apiResponse);
+    isLastPage = response.body['isLastPage'];
+    start = response.body['nextPageStart'] || -1;
+    pageCount++;
+  }
+  return values;
+};
+
+const sleepNow = (delay: number): unknown =>
+  new Promise((resolve) => setTimeout(resolve, delay));
+
+export async function listBitbucketServerProjects(
+  sourceUrl?: string,
+): Promise<BitbucketServerProjectData[]> {
+  const bitbucketServerToken = getBitbucketServerToken();
+  if (!sourceUrl) {
+    throw new Error(
+      'Please provide required `sourceUrl` for Bitbucket Server source',
+    );
+  }
+  debug(`Fetching all projects data`);
+  const projects = (await fetchAllProjects(
```

Comment:
we have to filter out "bad" repos, anything that is archived should not be here. Can you make sure to check for that?

Also can we pick the relevant data only please, not everything that BBS APi returns. See Github implementation where we map only the relevant fields that are a must for import, identification or matching later

Reply:
The resulting file has all this data: 
```
{"targets":[{"target":{"slug":"empty-repo","id":4114,"name":"empty-repo","scmId":"git","state":"AVAILABLE","statusMessage":"Available","forkable":true,"project":{"key":"EMPTY","id":323,"name":"Empty","public":false,"type":"NORMAL","links":{"self":[{"href":"http://bitbucket-server.example.io/projects/EMPTY"}]}},"public":false,"links":{"clone":[{"href":"http://bitbucket-server.example.io/scm/empty/empty-repo.git","name":"http"},{"href":"ssh://git@bitbucket-server.example.io:7999/empty/empty-repo.git","name":"ssh"}],"self":[{"href":"http://bitbucket-server.example.io/projects/EMPTY/repos/empty-repo/browse"}]}},"integrationId":"d9a2aab7-de07-458b-a724-561e7e68f5e2","orgId":"ad6a555b-90c8-44eb-8a3b-0a264f1292f9"}]}```

Can we trim this down only to must have data?

Reply:
Done